src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php

```diff
@@ -353,6 +353,12 @@ private function createFirewall(ContainerBuilder $container, string $id, array $
         $container->register($firewallEventDispatcherId, EventDispatcher::class)
             ->addTag('event_dispatcher.dispatcher', ['name' => $firewallEventDispatcherId]);
 
+        $container
+            ->getDefinition('security.firewall.event_dispatcher_map')
+            ->setLazy(true)
+            ->addMethodCall('addEventDispatcher', [$id, new Reference($firewallEventDispatcherId)])
+        ;
```

Comment:
This gets a bit tricky, but I would prefer if we used the [`ServiceLocator`](Symfony\Component\DependencyInjection\ServiceLocator) here instead of a lazy definition (which only works if a third party library is installed IIRC). We can then also remove the `FirewallEventDispatcherMapInterface` and `FirewallEventDispatcherMap` classes.

We use this strategy also for the `security.authenticator.managers_locator` service. Relevant code snippets are: https://github.com/symfony/symfony/blob/8bdbd3473ae3d51384a20cc3781971a833a4a5d1/src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php#L50-L51 https://github.com/symfony/symfony/blob/8bdbd3473ae3d51384a20cc3781971a833a4a5d1/src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php#L453-L454

Reply:
Thanks, fixed.